shiny/ui/_chat.py

```diff
@@ -1325,6 +1376,210 @@ async def _send_custom_message(self, handler: str, obj: ClientMessage | None):
+
+        session = get_current_session()
+        if session is None or isinstance(session, ExpressStubSession):
+            return BookmarkCancelCallback(lambda: None)
+
+        if session.bookmark.store == "disable":
+            raise ValueError(
+                "Bookmarking requires a `bookmark_store` to be set. Please set `bookmark_store=` in `App()`."
+            )
+
+        # Reset prior bookmarking hooks
+        self._destroy_bookmarks()
+
+        # Must use `root_session` as the id is already resolved. :-/
+        # Using a proxy session would double-encode the proxy-prefix
+        root_session = session.root_scope()
+        root_session.bookmark.exclude.append(self.id + "_user_input")
+
+        if bookmark_on is not None:
+
+            # When ever the bookmark is requested, update the query string (indep of store type)
+            @root_session.bookmark.on_bookmarked
+            async def _(url: str):
+                await session.bookmark.update_query_string(url)
+
+        if bookmark_on == "response":
+
+            @reactive.effect
+            # @reactive.event(lambda: chat.latest_message_stream.result, ignore_init=True)
+            @reactive.event(lambda: self.messages(format=MISSING), ignore_init=True)
+            async def _():
+                messages = self.messages(format=MISSING)
+
+                if len(messages) == 0:
+                    return
+
+                last_message = messages[-1]
+
+                if last_message.get("role") == "assistant":
+                    await session.bookmark()
+
+        @root_session.bookmark.on_bookmark
+        async def cancel_on_bookmark_client(state: BookmarkState):
+            if resolved_bookmark_id_str in state.values:
+                raise ValueError(
+                    f'Bookmark value with id (`"{resolved_bookmark_id_str}"`) already exists.'
+                )
+
+            with reactive.isolate():
+                state.values[resolved_bookmark_id_str] = await get_state()
+
+        @root_session.bookmark.on_bookmark
+        def cancel_on_bookmark_ui(state: BookmarkState):
+            if resolved_bookmark_id_msgs_str in state.values:
+                raise ValueError(
+                    f'Bookmark value with id (`"{resolved_bookmark_id_msgs_str}"`) already exists.'
+                )
+
+            with reactive.isolate():
+                # This only contains `ui.Chat(messages=)`
+                # This does NOT contain the `chat.ui(messages=)` values.
+                # When restoring, the `chat.ui(messages=)` values will need to be kept
+                # and the `ui.Chat(messages=)` values will need to be reset
+                state.values[resolved_bookmark_id_msgs_str] = self.messages(
+                    format=MISSING
+                )
+
+        ###############
+        # On Restore
+
+        # Attempt to stop the initialization of the `ui.Chat(messages=)` messages
+        self._init_chat.destroy()
```

Comment:
Should this only be done on restore?

Reply:
If we wait till the reactive graph is ready, I'd fear that we'd be too late. 

Where as keeping it here would cancel it as soon as possible